File: experiments/post_training/bio_tasks/generators/real_enrichment.py
# Copyright The Marin Authors
# SPDX-License-Identifier: Apache-2.0

"""Audit effect shrinkage and two enrichment specifications on observed RNA counts."""

import csv
import io
import json

from experiments.post_training.bio_tasks.contract import Contract, TableContract
from experiments.post_training.bio_tasks.generators.real_expression import integer, observations
from experiments.post_training.bio_tasks.generators.real_rnaseq import log_probability, quantity
from experiments.post_training.bio_tasks.real_data import mammary_samples, source_text, tsv_text
from experiments.post_training.bio_tasks.recipe_types import DataOrigin, Instance, OracleRuntime, Recipe, WorkflowScope

NAME = "real-rnaseq-shrinkage-enrichment-audit"
VIEWS = ("adjusted_tested", "raw_genome")
RESULT_COLUMNS = {
    "base_mean": "baseMean",
    "log2_fold_change": "log2FoldChange",
    "standard_error": "lfcSE",
    "wald_statistic": "stat",
    "neg_log10_p": "pvalue",
    "neg_log10_padj": "padj",
    "shrunken_log2_fold_change": "shrunken_log2FC",
    "shrunken_standard_error": "shrunken_lfcSE",
}


def reference_rows(name: str) -> list[dict[str, str]]:
    content = source_text("GSE60450", f"gse60450-luminal-shrinkage-{name}.tsv.gz")
    return list(csv.DictReader(io.StringIO(content), delimiter="\t"))


def generate_enrichment(_seed: int) -> Instance:
    metadata = sorted(mammary_samples(), key=lambda row: row["sample"])
    samples = [row["sample"] for row in metadata if row["population"] == "luminal"]
    names, genes, counts = observations()
    matrix = counts[:, [names.index(sample) for sample in samples]]
    totals = dict(zip(genes, map(int, matrix.sum(axis=1)), strict=True))
    fitted = {
        row["gene"]: {key: None if row[key] == "NA" else float(row[key]) for key in RESULT_COLUMNS.values()}
        for row in reference_rows("gene-results")
    }
    assert set(fitted) == {gene for gene, total in totals.items() if total >= 10}
    factors = {row["sample"]: float(row["size_factor"]) for row in reference_rows("size-factors")}
    raw = {gene for gene, row in fitted.items() if row["pvalue"] is not None and row["pvalue"] < 0.05}
    adjusted = {gene for gene, row in fitted.items() if row["padj"] is not None and row["padj"] < 0.05}
    large = {
        gene for gene, row in fitted.items() if row["shrunken_log2FC"] is not None and abs(row["shrunken_log2FC"]) >= 1
    }
    unshrunk_large = {
        gene for gene, row in fitted.items() if row["log2FoldChange"] is not None and abs(row["log2FoldChange"]) >= 1
    }
    tables = {
        "sample_qc.tsv": TableContract(
            columns={
                "library_counts": integer("counts before gene filtering", "counts"),
                "detected_genes": integer("genes with positive counts before filtering", "genes"),
                "size_factor": quantity("ratio size factor from filtered model", "factor"),
            },
            expected={
                sample: {
                    "library_counts": int(matrix[:, i].sum()),
                    "detected_genes": int((matrix[:, i] > 0).sum()),
                    "size_factor": factors[sample],
                }
                for i, sample in enumerate(samples)
            },
            max_bytes=16384,
        ),
        "gene_results.tsv": TableContract(
            columns={
                key: quantity(
                    "named coefficient " + value,
                    "log probability" if key.startswith("neg_log10") else "model quantity",
                    nullable=True,
                )
                for key, value in RESULT_COLUMNS.items()
            },
            expected={
                gene: {
                    key: log_probability(row[value]) if key.startswith("neg_log10") else row[value]
                    for key, value in RESULT_COLUMNS.items()
                }
                for gene, row in fitted.items()
            },
            max_bytes=16 * 1024 * 1024,
        ),
        "gene_decisions.tsv": TableContract(
            columns={
                "count_sum": integer("total over the six luminal libraries", "counts"),
                **{
                    name: integer("0/1 membership in the declared mask", "indicator")
                    for name in ("fitted", "raw_significant", "adjusted_significant", "large_shrunken_effect")
                },
            },
            expected={
                gene: {
                    "count_sum": total,
                    "fitted": int(gene in fitted),
                    "raw_significant": int(gene in raw),
                    "adjusted_significant": int(gene in adjusted),
                    "large_shrunken_effect": int(gene in large),
                }
                for gene, total in totals.items()
            },
            max_bytes=4 * 1024 * 1024,
        ),
    }
    backgrounds = {view: {row["gene"] for row in reference_rows(view + "-background")} for view in VIEWS}
    selections = {view: {row["gene"] for row in reference_rows(view + "-selected")} for view in VIEWS}
    tables["annotation_membership.tsv"] = TableContract(
        columns={
            view + "_" + kind: integer("0/1 membership after BP annotation intersection", "indicator")
            for view in VIEWS
            for kind in ("background", "selected")
        },
        expected={
            gene: (
                {view + "_background": int(gene in backgrounds[view]) for view in VIEWS}
                | {view + "_selected": int(gene in selections[view]) for view in VIEWS}
            )
            for gene in sorted(backgrounds["raw_genome"])
        },
        max_bytes=4 * 1024 * 1024,
    )
    significant = {}
    for view in VIEWS:
        rows = reference_rows(view + "-ora")
        significant[view] = {row["ID"] for row in rows if float(row["p.adjust"]) < 0.05}
        tables[view + "_enrichment.tsv"] = TableContract(
            columns={
                **{
                    key: integer(description, "genes")
                    for key, description in {
                        "overlap": "selected genes in the term",
                        "term_size": "term genes in the background",
                        "selected_genes": "annotated selected genes",
                        "background_genes": "annotated background genes",
                    }.items()
                },
                "neg_log10_p": quantity("upper-tail hypergeometric probability as capped -log10", "log probability"),
                "neg_log10_padj": quantity("native BH probability as capped -log10", "log probability"),
            },
            expected={
                row["ID"]: {
                    "overlap": int(row["Count"]),
                    "term_size": int(row["BgRatio"].split("/")[0]),
                    "selected_genes": len(selections[view]),
                    "background_genes": len(backgrounds[view]),
                    "neg_log10_p": log_probability(float(row["pvalue"])),
                    "neg_log10_padj": log_probability(float(row["p.adjust"])),
                }
                for row in rows
            },
            max_bytes=4 * 1024 * 1024,
        )
    summary = {
        "genes_fitted": len(fitted),
        "raw_significant": len(raw),
        "adjusted_significant": len(adjusted),
        "large_shrunken_effects": len(large),
        "changed_effect_threshold": len(large ^ unshrunk_large),
        "adjusted_tested_significant_terms": len(significant["adjusted_tested"]),
        "raw_genome_significant_terms": len(significant["raw_genome"]),
        "shared_significant_terms": len(significant["adjusted_tested"] & significant["raw_genome"]),
    }
    return Instance(
        "Audit the lactation-associated expression response in luminal mammary cells from GSE60450. "
        "The two requested enrichment specifications reflect different analysis choices. Determine their "
        "gene/term selections and quantify how coefficient shrinkage changes the absolute-effect threshold. "
        "Inputs are in /app/inputs. Join samples.tsv to counts.tsv by sample ID; Length is annotation. "
        "Use all six luminal libraries, with two biological replicates at each stage. Do not infer donor pairing. "
        "Retain genes with count sum >=10 across those libraries. Use DESeq2 1.50.2, design ~stage, "
        "levels virgin, '18.5 dP', '2 dL', ratio size factors, parametric dispersions, Wald tests, betaPrior=FALSE, "
        "and no count replacement. Keep pregnant libraries in the fit. Extract the named stage_2.dL_vs_virgin "
        "coefficient (not the character contrast interface, which resets genes zero in both compared stages), "
        "using alpha=.05, BH adjustment, independentFiltering=TRUE and cooksCutoff=TRUE. "
        "Apply apeglm 1.32.0 shrinkage to that coefficient with apeMethod=nbinomCR and the same Wald result. "
        "Preserve original p-values, BH values and Wald statistics; do not derive significance from shrunken effects. "
        "Write sample_qc.tsv using unfiltered library totals/detected genes and fitted size factors; "
        "gene_results.tsv for every fitted gene; and gene_decisions.tsv for all 27,179 input genes. "
        "Raw/adjusted significance is finite p/padj <.05; large_shrunken_effect means finite absolute shrunken "
        "log2FC >=1 regardless of significance. Excluded genes receive zero membership indicators. "
        "All probability fields are -log10(max(p,1e-300)); preserve NA model values. "
        "Use the installed org.Mm.eg.db 3.22.0 and GO.db 3.22.0, propagated GOALL biological-process memberships "
        "whose IDs are valid BP terms in GO.db. adjusted_tested selects adjusted-significant genes and uses "
        "finite-raw-p genes as background; raw_genome selects raw-significant genes and uses all genes in the "
        "annotation database with valid BP annotation as background. Intersect each selection/background with "
        "annotated genes before testing. Neither ORA selection includes an effect-size or directional cutoff. "
        "Write annotation_membership.tsv for every annotated gene, including genes absent from the count matrix. "
        "For each specification run clusterProfiler 4.18.4 enrichGO, keyType=ENTREZID, ont=BP, readable=FALSE, "
        "BH, minGSSize=10, maxGSSize=500, pvalueCutoff=qvalueCutoff=1; set seed 42 before each call. "
        "Export its complete @result hypothesis family to the corresponding enrichment table: size-eligible "
        "terms with nonzero selected overlap. Preserve the native BH denominator; do not append zero-overlap "
        "terms or apply Storey-qvalue filtering. Term significance is p.adjust <.05. "
        "Return an answer.json array containing one object with id=audit and all eight summary fields. "
        "changed_effect_threshold "
        "counts fitted genes whose absolute effect >=1 status differs between raw and shrunken coefficients. "
        "The two ORA specifications change selection and background jointly, so their difference cannot isolate "
        "one cause. Enrichment does not establish pathway activation or causal developmental effects.",
        {
            "counts.tsv": source_text("GSE60450", "gse60450-counts.txt.gz"),
            "samples.tsv": tsv_text(metadata),
            "query.json": (
                json.dumps({"analysis": NAME, "population": "luminal", "coefficient": "stage_2.dL_vs_virgin"}) + "\n"
            ),
        },
        Contract(
            columns={key: integer("count under the specified analysis rules", "genes or terms") for key in summary},
            expected={"audit": summary},
            tables=tables,
        ),
        {
            "raw_p_is_not_bh": [{"id": "audit", **summary, "adjusted_significant": len(raw)}],
            "unshrunk_is_not_shrunken": [{"id": "audit", **summary, "large_shrunken_effects": len(unshrunk_large)}],
        },
        data_origin=DataOrigin.REAL,
        source_ids=("GSE60450", "GO:mouse-3.22.0"),
        workflow_scope=WorkflowScope.CONNECTED,
        derivation=(
            "Complete deposited counts and sample identities; pinned annotation databases. Private native references "
            "have separate Python checks of count filters, normalized means, Wald tails, ORA sets and BH arithmetic."
        ),
    )


RECIPES = (
    Recipe(
        NAME,
        "1",
        (
            "replicate-aware count models",
            "effect shrinkage",
            "multiple testing",
            "annotation provenance",
            "enrichment universe sensitivity",
        ),
        ("TSV count matrix", "TSV sample metadata", "Bioconductor annotation databases"),
        (
            "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE60450",
            "https://bioconductor.org/packages/DESeq2",
            "https://bioconductor.org/packages/clusterProfiler",
        ),
        generate_enrichment,
        oracle_timeout=1800,
        oracle_runtime=OracleRuntime.NATIVE,
    ),
)

File: experiments/post_training/bio_tasks/solvers/real_enrichment.R
# Copyright The Marin Authors
# SPDX-License-Identifier: Apache-2.0

# Fresh input-reading oracle; package outputs are checked against separately prepared references.
suppressPackageStartupMessages(library(DESeq2))
suppressPackageStartupMessages(library(apeglm))
suppressPackageStartupMessages(library(clusterProfiler))
suppressPackageStartupMessages(library(org.Mm.eg.db))
suppressPackageStartupMessages(library(GO.db))
args <- commandArgs(TRUE)
stopifnot(length(args) == 2)
input <- args[1]; output <- args[2]
versions <- c(DESeq2="1.50.2",apeglm="1.32.0",clusterProfiler="4.18.4",org.Mm.eg.db="3.22.0",GO.db="3.22.0")
for (name in names(versions)) stopifnot(as.character(packageVersion(name)) == versions[[name]])
options(digits=17, enrichment_force_universe=FALSE)
write_artifact <- function(rows, name) {
  write.table(rows, file.path(output,name), sep="\t", quote=FALSE, row.names=FALSE, na="NA")
}
logp <- function(values) -log10(pmax(values,1e-300))
counts <- read.delim(file.path(input,"counts.tsv"),row.names=1,check.names=FALSE)
metadata <- read.delim(file.path(input,"samples.tsv"),stringsAsFactors=FALSE)
samples <- metadata[metadata$population == "luminal",,drop=FALSE]
samples <- samples[order(samples$sample),,drop=FALSE]
rownames(samples) <- samples$sample
samples$stage <- factor(samples$stage,levels=c("virgin","18.5 dP","2 dL"))
matrix <- as.matrix(counts[,samples$sample,drop=FALSE])
stopifnot(!anyDuplicated(rownames(matrix)),all(matrix >= 0),all(matrix == floor(matrix)),all(table(samples$stage)==2))
storage.mode(matrix) <- "integer"
keep <- rowSums(matrix) >= 10
dds <- DESeqDataSetFromMatrix(matrix[keep,,drop=FALSE],samples,~stage)
dds <- estimateSizeFactors(dds,type="ratio")
dds <- estimateDispersions(dds,fitType="parametric",quiet=TRUE)
dds <- nbinomWaldTest(dds,betaPrior=FALSE,quiet=TRUE)
coefficient <- "stage_2.dL_vs_virgin"
stopifnot(coefficient %in% resultsNames(dds))
result <- results(dds,name=coefficient,alpha=.05,pAdjustMethod="BH",independentFiltering=TRUE,cooksCutoff=TRUE)
shrunk <- lfcShrink(dds,coef=coefficient,res=result,type="apeglm",apeMethod="nbinomCR",parallel=FALSE)
stopifnot(identical(rownames(result),rownames(shrunk)))
write_artifact(data.frame(id=colnames(matrix),library_counts=colSums(matrix),detected_genes=colSums(matrix>0),
                         size_factor=unname(sizeFactors(dds))),"sample_qc.tsv")
write_artifact(data.frame(id=rownames(result),base_mean=result$baseMean,log2_fold_change=result$log2FoldChange,
                         standard_error=result$lfcSE,wald_statistic=result$stat,neg_log10_p=logp(result$pvalue),
                         neg_log10_padj=logp(result$padj),shrunken_log2_fold_change=shrunk$log2FoldChange,
                         shrunken_standard_error=shrunk$lfcSE),"gene_results.tsv")
gene_ids <- rownames(result)
raw <- gene_ids[is.finite(result$pvalue) & result$pvalue < .05]
adjusted <- gene_ids[is.finite(result$padj) & result$padj < .05]
large <- gene_ids[is.finite(shrunk$log2FoldChange) & abs(shrunk$log2FoldChange) >= 1]
unshrunk_large <- gene_ids[is.finite(result$log2FoldChange) & abs(result$log2FoldChange) >= 1]
write_artifact(data.frame(id=rownames(matrix),count_sum=rowSums(matrix),fitted=as.integer(keep),
                         raw_significant=as.integer(rownames(matrix) %in% raw),
                         adjusted_significant=as.integer(rownames(matrix) %in% adjusted),
                         large_shrunken_effect=as.integer(rownames(matrix) %in% large)),"gene_decisions.tsv")
annotation <- AnnotationDbi::select(org.Mm.eg.db,keys=keys(org.Mm.eg.db,keytype="ENTREZID"),
                                  keytype="ENTREZID",columns=c("GOALL","ONTOLOGYALL"))
bp <- AnnotationDbi::Ontology(GOTERM)
valid <- names(bp)[bp == "BP"]
annotated <- sort(unique(as.character(annotation$ENTREZID[!is.na(annotation$GOALL) &
                  annotation$ONTOLOGYALL == "BP" & annotation$GOALL %in% valid])))
backgrounds <- list(adjusted_tested=intersect(gene_ids[is.finite(result$pvalue)],annotated),raw_genome=annotated)
selections <- list(adjusted_tested=intersect(adjusted,backgrounds$adjusted_tested),raw_genome=intersect(raw,annotated))
membership <- data.frame(id=annotated)
significant <- list()
for (view in names(backgrounds)) {
  universe <- backgrounds[[view]]; selected <- selections[[view]]
  membership[[paste0(view,"_background")]] <- as.integer(annotated %in% universe)
  membership[[paste0(view,"_selected")]] <- as.integer(annotated %in% selected)
  set.seed(42)
  enriched <- enrichGO(gene=selected,OrgDb=org.Mm.eg.db,keyType="ENTREZID",ont="BP",universe=universe,
                      pAdjustMethod="BH",minGSSize=10,maxGSSize=500,pvalueCutoff=1,qvalueCutoff=1,readable=FALSE)
  native <- enriched@result
  significant[[view]] <- native$ID[is.finite(native$p.adjust) & native$p.adjust < .05]
  term_size <- as.integer(sub("/.*","",native$BgRatio))
  write_artifact(data.frame(id=native$ID,overlap=native$Count,term_size=term_size,selected_genes=length(selected),
                           background_genes=length(universe),neg_log10_p=logp(native$pvalue),
                           neg_log10_padj=logp(native$p.adjust)),paste0(view,"_enrichment.tsv"))
}
write_artifact(membership,"annotation_membership.tsv")
answer <- list(list(id="audit",genes_fitted=nrow(result),raw_significant=length(raw),adjusted_significant=length(adjusted),
                   large_shrunken_effects=length(large),
                   changed_effect_threshold=length(setdiff(union(large,unshrunk_large),intersect(large,unshrunk_large))),
                   adjusted_tested_significant_terms=length(significant$adjusted_tested),
                   raw_genome_significant_terms=length(significant$raw_genome),
                   shared_significant_terms=length(intersect(significant$adjusted_tested,significant$raw_genome))))
jsonlite::write_json(answer,file.path(output,"answer.json"),auto_unbox=TRUE,pretty=TRUE,na="null",digits=16)
